core/services/ocr2/plugins/ccip/ccipexec/batching.go

```
@@ -50,6 +50,11 @@ func buildExecutionReportForMessages(
 			continue
 		}
 		innerIdx := int(observedMessage.SeqNr - commitInterval.Min)
+		if innerIdx >= len(msgsInRoot) || innerIdx < 0 {
+			return ccipdata.ExecReport{}, fmt.Errorf("invalid inneridx SeqNr=%d IntervalMin=%d msgsInRoot=%d",
+				observedMessage.SeqNr, commitInterval.Min, len(msgsInRoot))
+		}
```

Is this an invariant violation that needs to be investigated further? 

Also is this something we should return an error from and we're guaranteed to make progress? Or should we put a `continue` instead?

---

I see that the call site passes in a slice that can be too small, since this is being called in the `sort.Search` context which does a binary search.

---

> Is this an invariant violation that needs to be investigated further?
yes

I believe we don't want to make progress in that case, something seems corrupted.
@connorwstein wdyt?

---

We've briefly discussed this with @makramkd and it seems that the only scenario in which this could happen is when nodes see different states of the chain. For instance, the node trying to execute that Commit Root doesn't have all OnRamp messages from that root in the database. It can happen when a node has slow RPC on a fast chain when the time between publishing a message to OnRamp and creating a Commit Report is very low. Majority of nodes agree on observed messages (and min/max sequence number) and then slow nodes can't get all the OnRamp messages from the database.

If that's true then it would be better to have this check closer to the place responsible for fetching onramp messages, e.g.
```go
func getProofData(
	ctx context.Context,
	sourceReader ccipdata.OnRampReader,
	interval ccipdata.CommitStoreInterval,
) (sendReqsInRoot []ccipdata.Event[internal.EVM2EVMMessage], leaves [][32]byte, tree *merklemulti.Tree[[32]byte], err error) {
	
	sendReqs, err := sourceReader.GetSendRequestsBetweenSeqNums(ctx, interval.Min, interval.Max, false)
	if len(sendReqs) != Max - Min + 1 {
      I still don't have all onramp messages! ! ! 
    }
```
@dimkouv @connorwstein

---

We can consider adding that check inside `sourceReader.GetSendRequestsBetweenSeqNums`

---

yeah that makes sense, but let's keep the plugin validations to prevent the panic as an extra safety measure

---

Yeah, we want plugin to be defensive, but let's add extra check to reader. We should be able to write a simple test for when implemented in reader ;)

---

Actually had a second thought while starting to work on this this seems better the way it is right now.

1. We have multiple implementations of that `GetSendRequestsBetweenSeqNums` so we will end up having the exact same validation duplicated on each implementation.
2. The API of `GetSendRequestsBetweenSeqNums` does not guarantee that all the messages exist, what it does is fetching messages inside the provided interval, so it is the job of the caller to validate the response.
3. If we move the validation inside the implementation it is easy for someone that adds a new implementation to forget adding that validations.
4. We have more parts of the code that call `GetSendRequestsBetweenSeqNums` which I am not sure they require all the seq nrs to be there.

---

> We have multiple implementations of that GetSendRequestsBetweenSeqNums so we will end up having the exact same validation duplicated on each implementation.

That's the flaw of the current versioning implementation, we already have a lot of code duplication.

> If we move the validation inside the implementation it is easy for someone that adds a new implementation to forget adding that validations.

That's true, but this problem is present for all readers having more "advanced" methods 

> We have more parts of the code that call GetSendRequestsBetweenSeqNums which I am not sure they require all the seq nrs to be there.

If that's the case then I would still value this check to be closer to the `GetSendRequestsBetweenSeqNums` call. Otherwise, it might be hard for the reader to understand why this doesn't fail fast. Also I would suggest adding a comment describing scenario in which it could fail, because it's not that obvious, especially for new people

---

@dimkouv @makramkd any thoughts? It would be great to merge that today

---

added a comment